src/mongoc/mc_topology.erl

```diff
@@ -115,7 +115,7 @@ get_pool(Topology, Options) ->
 get_pool(RPMode, RPTags, State) ->
   TO = State#topology_state.topology_opts,
   ServerSelectionTimeoutMS = mc_utils:get_value(serverSelectionTimeoutMS, TO, 30000),
-  Pid = spawn(?MODULE, get_pool, [self(), State, RPMode, RPTags]),
+  Pid = proc_lib:spawn_link(?MODULE, get_pool, [self(), State, RPMode, RPTags]),
```

Comment:
here is a safe middleman delegation which prevents process leaking without contaminating caller's mailbox with EXIT or DOWN messages.
```
delegate(Fun) when is_function(Fun, 0) ->
    Caller = self(),
    Ref = make_ref(),
    {MiddleMan, Ref} =
    erlang:spawn_monitor(
      fun() ->
              process_flag(trap_exit, true),
              PRef = erlang:monitor(process, Caller),
              Worker = erlang:spawn_link(fun() -> Caller ! {Ref, Fun()} end),
              receive
                  {'DOWN', PRef, process, _Pid, _Reason} ->
                      %% Caller is dead, no reason to continue
                      exit(Worker, kill),
                      exit(normal);
                  {'EXIT', Worker, Reason} ->
                      %% worker has finished its job
                      exit(Reason)
              end
      end),
    receive
        {Ref, Result} ->
            ok = demon(Ref),
            Result;
        {'DOWN', Ref, process, MiddleMan, Reason} ->
            %% propagate worker's exception back to the caller
            exit(Reason)
    end.

%% Optimized demonitor comparing to demonitor(Ref, [flush])
%% because it only tries to receive the 'DOWN' message
%% when the DOWN message is actually already sent
demon(Ref) ->
    case erlang:demonitor(Ref, [info]) of
        true ->
            ok;
        _ ->
            receive
                {'DOWN', Ref, process, _Pid, _Reason} ->
                    ok
            after
                0 ->
                    ok
            end
    end.
```

Reply:
a refined version.
NOTE: not tested!
```
delegate(Fun) when is_function(Fun, 0) ->
    Caller = self(),
    Ref = make_ref(),
    {MiddleMan, Ref} =
    erlang:spawn_monitor(
      fun() ->
              process_flag(trap_exit, true),
              PRef = erlang:monitor(process, Caller),
              %% Send the result straight to caller, but not through middleman.
              Worker = erlang:spawn_link(
                         fun() ->
                                 R = try
                                         {normal, Fun()}
                                     catch
                                         C : E : S ->
                                             {excep, {C, E, S}}
                                     end,
                                 Caller ! {Ref, R}

                         end),
              receive
                  {'DOWN', PRef, process, _Pid, _Reason} ->
                      %% Caller is dead, no reason to continue
                      exit(Worker, kill),
                      exit(normal);
                  {'EXIT', Worker, Reason} ->
                      %% worker has finished its job (Reason=normal)
                      %% or might have crashed due to exception when evaluating Fun
                      %% Propagate this Reason to Caller
                      exit(Reason)
              end
      end),
    receive
        {Ref, {normal, Result}} ->
            ok = demon(Ref),
            Result;
        {Ref, {excep, {C, E, S}}} ->
            ok = demon(Ref),
            erlang:raise(C, E, S);
        {'DOWN', Ref, process, MiddleMan, Reason} ->
            %% Although the worker is catching everything,
            %% there is still a chance that inside Fun()
            %% it link or spwan_link to another process which
            %% brings down the worker process.
            %% Propagate worker's exception back to the caller,
            %% and the caller may choose to catch this exit/1
            exit(Reason)
```

Reply:
Thanks for the reminder, using spawn_link was a huge mistake!
The role of spawn_link is bidirectional. Under normal circumstances, the child process returns quickly and ends. This will cause the parent process to be terminated as well.